runtimes/alibi-detect/mlserver_alibi_detect/runtime.py

```diff
@@ -105,23 +118,37 @@ def _detect(self, payload: InferenceRequest) -> InferenceResponse:
         input_data = self.decode_request(payload, default_codec=NumpyRequestCodec)
         predict_kwargs = self._ad_settings.predict_parameters
 
-        try:
-            y = self._model.predict(np.array(input_data), **predict_kwargs)
-            return self._encode_response(y)
-        except (ValueError, IndexError) as e:
-            raise InferenceError(
-                f"Invalid predict parameters for model {self._settings.name}: {e}"
-            ) from e
+        # If batch is configured, wrap X in a list so that it is not unpacked
+        X = np.array(input_data)
+        if self._ad_settings.batch_size:
+            X = [X]
+
+        # Run detector inference
+        pred = []
+        for x in X:
+            # Prediction
+            try:
+                pred.append(self._model.predict(x, **predict_kwargs))
+            except (ValueError, IndexError) as e:
+                raise InferenceError(
+                    f"Invalid predict parameters for model {self._settings.name}: {e}"
+                ) from e
+            # Save state if necessary
+            if self._ad_settings.state_save_freq and \
+                    self._model.t % self._ad_settings.state_save_freq == 0 and self._model.t > 0:
+                self._model.save_state(os.path.join(self._model_uri, 'state'))
```

Comment:
Could nest the saving logic in a `AlibiDetectRuntime._save_state` method if we want to raise custom errors etc

Reply:
Yuup! That would make sense - we can also re-use that method in the (future) `unload()` handler. We could also move the `if` check to a `_should_save_state()` method to keep the logic cleaner.

Reply:
Refactored in https://github.com/SeldonIO/MLServer/pull/1108/commits/4790d5f8c49aa1d49d993f5ff92bb2d0740b8435. Note that I haven't added  unit test to check that `_save_state` raises an error, since it is difficult to cause an error! Alibi Detect automatically creates a directory, so a `FileNotFoundError` isn't possible. I was thinking about forcing a `PermissionError` but it doesn't look easy to change the permissions of the pytest `tmp_path`...